impress-org/givewp
# feat: improve legacy stripe checkout using stripe elements #4775

Resolves #4775 

## Description
The most popular feature of Stripe Checkout is their modal popup which will get deprecated soon in replacement for Stripe Elements and Redirect method. As we have noticed that donors don't like getting redirected to a payment page for donation as it can decrease the conversions. So, as suggested by Stripe team, we have come up with a custom solution where we will open a custom modal and inside that modal we will load the credit card fields, Google Pay, and Apple Pay using the Stripe Elements to provide the donors with a choice of modal popup and the good thing is we will still call it Stripe Checkout with modal 🎉 

**With Legacy Stripe Checkout (Modal option):** We were able to accept donations via Credit Card only.
**With Stripe Checkout using Elements in Modal:** We are supporting Google Pay and Apple Pay along with Credit Card fields _(CC fields has two variations - Single Line Fields and Multi Line Fields)_

### Regarding this PR
**Default - Stripe Checkout Modal Design**
![image](https://user-images.githubusercontent.com/1852711/88793754-fa332700-d1ba-11ea-8631-6199b54856f8.png)

**With Background Image - Stripe Checkout Modal Design**
![image](https://user-images.githubusercontent.com/1852711/88793925-41b9b300-d1bb-11ea-9468-9bc7502cc304.png)

**P.S.:** In the above screenshots, I have loaded custom fonts due to which you are seeing different fonts loaded for Card Number, CVC, and Card Expiry fields.

## Affects
Introduced Stripe Checkout with Modal using Stripe Elements in replacement to Legacy Stripe Checkout.

## Pre-review Checklist

## Testing Instructions
1. Enable Stripe Checkout as payment gateway
2. Go to Stripe Checkout Settings under `Settings > Payment Gateways > Stripe > Stripe Checkout` and select **Checkout Type** to **Modal**
3. Enable Stripe Premium add-on to show Google and Apple Pay inside the Stripe Checkout Modal. Check PR: 
4. Try processing donation with Google Pay, Apple Pay **(if you can)**, and Credit Card fields **(Single and Multi fields)** of Stripe Checkout
5. Try processing donations from different variations of donation forms like Legacy Donation Form - with modal, reveal, and other supported types, Sequoia Form - New Form Template, Form Grid, Same Donation Form loaded on the same page.

## User Documentation
Need to update the existing documentation https://stripe.com/docs/legacy-checkout or need to create a separate one for the new Stripe Checkout modal functionality.
